modules/rtp.io/rtp_io.c

```diff
@@ -0,0 +1,254 @@
+    0,               /* exported async functions */
+    params,      	 /* param exports */
+    0,       	 /* exported statistics */
+    0,         	 /* exported MI functions */
+    NULL,       	 /* exported pseudo-variables */
+    0,		 /* exported transformations */
+    0,               /* extra processes */
+    0,        	 /* module pre-initialization function */
+    mod_init,        /* module initialization function */
+    0,               /* reply processing function */
+    mod_destroy,
+    child_init,      /* per-child init function */
+    0
+};
+
+#if defined(__linux__)
+static int optreset; /* Not present in linux */
+#endif
+
+static struct opt_save {
+    char *optarg;
+    int optind;
+    int optopt;
+    int opterr;
+    int optreset;
+} opt_save = {.optind = 1};
+
+#define OPT_SAVE() (opt_save = (struct opt_save){optarg, optind, optopt, opterr, optreset})
+#define OPT_RESTORE() ({ \
+    optarg = opt_save.optarg; \
+    optind = opt_save.optind; \
+    optopt = opt_save.optopt; \
+    opterr = opt_save.opterr; \
+    optreset = opt_save.optreset; \
+})
+
+#define howmany(x, y) (sizeof(x) / sizeof(y))
+
+static struct rtpp_env argv0 = {.cp = "rtpproxy"};
+
+static struct rtp_io_desc rpi_desc = {
+    .env = {.len = 1, .first = &argv0, .last = &argv0},
+};
+
+struct rtp_io_desc *rpi_descp = &rpi_desc;
+
+#define ENV_ADD(x, elabel, ...) { \
+    struct rtpp_env *_e = rtp_io_env_asprintf((x), ##__VA_ARGS__); \
+    if (_e == NULL) \
+        goto elabel; \
+    rtp_io_env_append(&rpi_descp->env, _e); \
+}
+
+static int
+rio_socks_init(int nsocks)
+{
+    size_t asize;
+    struct rtp_io_socks *socks;
+
+    LM_DBG("allocating %s(%d)\n", exports.name, nsocks);
+
+    asize = sizeof(struct rtp_io_socks) + (nsocks * sizeof(int) * 2);
+    socks = malloc(asize);
+    if (socks == NULL)
+        goto e0;
+
+    memset(socks, '\0', asize);
+    rpi_descp->socks = socks;
+    return 0;
+e0:
+    return -1;
+}
+
+static int mod_init(void)
+{
+    int nsocks;
+    const char * const argv_stat[] = {
+        "-d", "err",
+        "-n", "tcp:127.0.0.1:9642",
+        "--dso", "catch_dtmf",
+        "--dso", "dtls_gw",
+        "--dso", "ice_lite",
+    };
+
+    nsocks = count_child_processes();
+
+    if (nsocks <= 1)
+        goto e0;
+
+    LM_DBG("initializing %s(%d)\n", exports.name, nsocks);
+
+    if (rio_socks_init(nsocks) != 0)
+        goto e0;
+
+    for (int i = 0; i < howmany(argv_stat, *argv_stat); i++) {
+        ENV_ADD(argv_stat[i], e1);
+    }
+
+    for (int i = 0; i < nsocks; i++) {
+        int *fdp = &rpi_descp->socks->holder[i * 2];
+        if (socketpair(AF_UNIX, SOCK_STREAM, 0, fdp) < 0)
+            goto e1;
+        ENV_ADD("-s", e1);
+        ENV_ADD("fd:%d", e1, fdp[0]);
+    }
+
+
+    int argc;
+    const char *const *argv = rtp_io_env_gen_argv(&rpi_descp->env, &argc);
+    if (argv == NULL)
+        goto e1;
+
+    OPT_RESTORE();
+    rpi_descp->rtpp_cfsp = rtpp_main(argc, argv);
```

Comment:
This `rtpp_main()` function call is the entrypoint into the `librtpproxy.a` embedded RTPProxy daemon.  The default logic proceeds to fork `12` worker threads which will be part of the "Attendant" OpenSIPS worker.

This logic needs to be moved as part of a `proc_export_t` extra process of the module, to prevent any scenario where a thread aborts and prevents OpenSIPS shutdown cleanup procedure.  I hate to be pedantic here, but not only there are `188` modules following this rule and we shouldn't start making exceptions, but also it is very likely that `rtp.io` will gain a good amount of traction on release, which further reinforces the need of a clean design from the start.  This is not a feature request as I see it, it is a design requirement.

Cheers,

Reply:
OK, the threads have been moved into a dedicated extra process, please check. Thanks!